Feature/jv fda rest reader

sgbalogh opened the pull request:
This branch still has work to be done (most significantly, it's lacking tests), but wanted to make a PR just to get some initial comments.

I created a [resource set](https://github.com/NYULibraries/ichabod/blob/feature/jv_fda_rest_reader/app/resource_sets/vinopal_fda_collection.rb) for JV's FDA collection, which uses a new [FDA source reader](https://github.com/NYULibraries/ichabod/blob/feature/jv_fda_rest_reader/lib/ichabod/resource_set/source_readers/fda_collection_rest_reader.rb) that I also added, which connects to the FDA via the new DSpace REST API. My hope is that the source reader is generic enough to be used for a variety of FDA collections. In fact, the only unique information that the JV resource set provides to the reader is the corresponding FDA collection ID.

Here are some questions I have:
- If we would want to use the source reader for a variety of FDA collections, are there other considerations that might make it more versatile? For instance, a way for a resource set specification to provide its own arbitrary mapping of FDA fields into NYUCore?
- I'm using the ruby Rest-Client gem to handle HTTP connections; should I be using something else, or is that ok? I just noticed that the NYU Press source reader uses Faraday.
- A comment: When constructing parameters for the HTTP connections made, I had to set `:verify_ssl => false`, simply because my machines kept complaining about SSL certificate trust issues whenever that was not the case. Has anyone else run into this problem? I need to research it a bit more (and see if it has to do with my OpenSSL or if it has to do with archive.nyu.edu) –– but I acknowledge that we will definitely want to verify ssl when this is actually being used.

I also realized that I don't understand the concept of `editor` being used in Hydra. I'm assuming this has something to do with user access levels?

Any comments would be very much appreciated! This has been a really good experiment for me to learn a bit about how Hydra works, but I have a lot more to learn...

da70 commented:
The SSL cert issue sounds familiar.  An error like that had arisen recently when loading some FDA data and Kate figured out what the problem was (it wasn't actually the certificate).  For a while we had to comment out an FDA data-load in `spec_helper.rb` to get around it until Kate replaced it with load from CSV (I believe).  I don't actually know the details.

eshadatta commented on lib/ichabod/resource_set/source_readers/fda_collection_rest_reader.rb:
```diff
@@ -0,0 +1,116 @@
+        def available_link_from_record(record)
+          values = []
+          record['metadata'].each do |element|
+            if element['key'].include? "dc.identifier"
```
@sgbalogh ,

Could you make this a one line conditional statement? 
```
if element['key'].include? "dc.identifier" && element['value'].include? "handle"
        values << element['value']
    end
```
 unless I'm missing something?

eshadatta commented:
@sgbalogh ,

I noticed that you're calling the ENV var directly, examples: [1](https://github.com/NYULibraries/ichabod/blob/ced175c0931c2b317e6a93ef463f1572096aebff/lib/ichabod/resource_set/source_readers/fda_collection_rest_reader.rb#L67), [2](https://github.com/NYULibraries/ichabod/blob/ced175c0931c2b317e6a93ef463f1572096aebff/lib/ichabod/resource_set/source_readers/fda_collection_rest_reader.rb#L86), and [3](https://github.com/NYULibraries/ichabod/blob/ced175c0931c2b317e6a93ef463f1572096aebff/lib/ichabod/resource_set/source_readers/fda_collection_rest_reader.rb#L102). I would suggest making an attribute reader, fda_rest_url, grabbing the parameter from args and assigning them in [deploy.rb](https://github.com/NYULibraries/ichabod/blob/development/config/deploy.rb) passing the ENV var as an argument. In the [resource_set](https://github.com/NYULibraries/ichabod/blob/ced175c0931c2b317e6a93ef463f1572096aebff/app/resource_sets/vinopal_fda_collection.rb) for this collection, pass the ENV var as an argument. 

So, for example, in fda_collection.rb:
```
attr_reader :fda_collection_id, :fda_rest_url
def initialize(*args)
    @fda_collection_id = '520'
    @fda_rest_url = args.shift
    super
  end
```

in the source reader:
```
def_delegators :resource_set, :fda_collection_id, :fda_rest_url
```
and call fda_rest_url instead of the ENV var.

This is just the style of how these are written in ichabod and also in case the ENV var name is changed, you just need to change it in deploy.rb.

The same thing with [this](https://github.com/NYULibraries/ichabod/blob/ced175c0931c2b317e6a93ef463f1572096aebff/lib/ichabod/resource_set/source_readers/fda_collection_rest_reader.rb#L91) ENV var.

Great start on this! @NYULibraries/hydra , please feel free to answer his questions or have feedback for him. In terms of your questions, Stephen, I'll try to answer some of them, but it might be better to discuss them as a group.

I'm copying one question, so I'll answer it inline.
- If we would want to use the source reader for a variety of FDA collections, are there other considerations that might make it more versatile? For instance, a way for a resource set specification to provide its own arbitrary mapping of FDA fields into NYUCore?
  - We could probably send a hash but my recommendation is to wait till we have more use cases to iterate on a more standardized approach. Of course, others may disagree.

Hope this helps.

Thanks.

sgbalogh commented:
@eshadatta: thanks so much for these comments! This is incredibly helpful. I'll have a block of time to dedicate to implementing these changes tomorrow.

-s

da70 commented:
Question for anybody: I notice that generally we hardcode the URLs in `deploy.rb`, the one exception being `spatial_data_repository`, which appears to use `ENV` var for sensitive user token.  Does Stephen's work need to be using `ENV`, or can he not just hardcode the value that would be assigned to `fda_rest_url`?  Just wondering what the criteria are for choosing one or the other.

eshadatta commented:
@da70 ,

Most of the urls are using env vars in [deploy.rb](https://github.com/NYULibraries/ichabod/blob/development/config/deploy.rb). Examples: [1](https://github.com/NYULibraries/ichabod/blob/development/config/deploy.rb#L24), [2](https://github.com/NYULibraries/ichabod/blob/development/config/deploy.rb#L39), [3](https://github.com/NYULibraries/ichabod/blob/development/config/deploy.rb#L39), [4](https://github.com/NYULibraries/ichabod/blob/development/config/deploy.rb#L39) and there are more. They are assigned in the task :set_variables.

da70 commented:
@sgbalogh,
Oh sorry, I happened to glance at `deploy.rb` on my computer and saw this:
https://github.com/NYULibraries/ichabod/blob/development/config/deploy.rb#L114-139

I don't even know what that `:ingest` namespace is for...still learning my way around.

sgbalogh commented:
@eshadatta: I implemented the suggestions you made (thanks again!), adding attribute accessors for the API endpoint, as well as for user/pass. I also added Capistrano deployment instructions/vars to deploy.rb, though I haven't used Capistrano in the past, so not 100% sure that I configured it correctly. For instance: do I need to register the namespace (in this case, I used `fda_jv`) somewhere other than on deploy.rb, or is that only a concern at the Jenkins/CI level (therefore, getting Jenkins to actually use my FDA deploy command periodically)?

I still have some lingering confusions about writing the tests, but hopefully I can talk to you about that tomorrow or when we meet on Monday.

da70 commented:
@sgbalogh, I notice that in `spec_helper.rb` you added to `VCR.configure` at the end these filters:

```ruby
  c.filter_sensitive_data('user') { ENV['FDA_REST_USER'] }
  c.filter_sensitive_data('password') { ENV['FDA_REST_PASS'] }
```

These would seem to overwrite the ones already added for Rosie a few lines up:

```ruby
  c.filter_sensitive_data('user') { ENV['ICHABOD_ROSIE_USER'] }
  c.filter_sensitive_data('password') { ENV['ICHABOD_ROSIE_PASSWORD'] }
```

da70 commented:
@sgbalogh 

To be more precise, those 2 new `filter_sensitive_data` calls add 2 new `VCR::Hooks::FilteredHook` instances to the `before_playback` and `before_record` hook arrays, which presumably run after the Rosie hooks callbacks.

So now that I think about it, this would explain why Rosie doesn't fail.  The Rosie filters run first and replace the placeholder strings "user" and "password".  By the time your filters run the placeholders are gone.

This would suggest that your specs should fail when running via VCR (assuming the credentials are different from Rosie).  On my local they don't fail, though.

da70 commented:
On my local I commented out all `c.filter_sensitive_data(...)` calls (for both this PR and for Rosie, just to be thorough), and ran `fda_rest_reader_spec.rb`.  All 10 tests passed, so we could probably remove these lines from the VCR configuration.

Was talking to Kate and she said the collection for this user story is public, so wouldn't need credentials.
This which makes the 403 that Travis is getting back perplexing (assuming it's making a live request).

eshadatta commented:
@sgbalogh @da70 
Just to add, we need the filter_sensitive data when making the cassette. One can also change user and password to be whatever you want, if you want to call it fda_user and fda_password and see if they all pass since user and password is being used for rosie. In either case, the filter_sensitive data is needed when first making a cassette. But if it's public and one doesn't need to login, then perhaps the point is moot. Thanks.

da70 commented:
@eshadatta, ah I see, on cassette record.  Hadn't thought of that.

da70 commented:
Stephen and I are looking at `VCR.http_interactions.interactions` array in scope of his new `fda_authenticate` method.  It looks like the Rosie filters have mangled it:

```
#<struct VCR::HTTPInteraction request=#<struct VCR::Request method=:post, uri="https://archive.nyu.edu/rest/login", body="{\"email\":\"[REDACTED MANGLED USERNAME]\",\"[REDACTED ROSIE PASSWORD THAT REPLACED PARAM KEY NAME]\":\"[REDACTED ROSIE PASSWORD THAT REPLACED FDA PARAM VALUE]\"}", headers={"Accept"=>["*/*; q=0.5, application/xml"], "Accept-Encoding"=>["gzip, deflate"], "Content-Type"=>["application/json"], "Content-Length"=>["53"], "User-Agent"=>["Ruby"]}>, response=#<struct VCR::Response status=#<struct VCR::ResponseStatus code=200, message="OK">, headers={"Date"=>["Thu, 30 Jun 2016 19:40:43 GMT"], "Server"=>["Apache-Coyote/1.1"], "Content-Type"=>["text/plain; charset=UTF-8"], "Transfer-Encoding"=>["chunked"]}, body="77962713-538e-4fae-a222-8b3d01abf550", http_version=nil, adapter_metadata={}>, recorded_at=2016-06-30 15:40:42 -0400>
```

* [REDACTED MANGLED USERNAME]: this is an interesting mangling case...the Rosie filter replaced the `email` value with a username that then subsequently got mangled by the new FDA filter because the Rosie username has the word "user" in it.  The result was a username that is not valid for either.
* [REDACTED ROSIE PASSWORD THAT REPLACED PARAM KEY NAME]: the Rosie filter replaced the POST parameter name "password" with the actual Rosie password.  This is because `filter_sensitive_data` creates a filter that does string replacements regardless of where the string occurs.
* [REDACTED ROSIE PASSWORD THAT REPLACED FDA PARAM VALUE]: the Rosie filter replaced Stephen's placeholder for the password value.

We should probably make sure that placeholder strings are very unique to prevent unwanted replacements in the wrong context.  Both Rosie and FDA should probably change their placeholder strings to something unusual enough to not only keep them distinct from each other, but to make it unlikely that they will be found anywhere else like the response body or GET/POST param key names.

We still don't know why this test succeeds on local machines but not on Travis.

da70 commented:
Stephen and I tested to see if Travis build of his branch was actually hitting dev-dl-pa for Rosie tests.  I `tail -f` the logs and there were no requests from Travis when we refreshed the build.

Very odd, because when we ran the Rosie tests on my machine all the requests showed up in the dev-dl-pa server apache logs.  This is because the FDA VCR filters were mangling the Rosie cassette interactions, causing VCR to treat them as brand new interactions to be recorded (and we verified using git that the cassette had been updated with new interactions).  The same thing should have happened in the Travis build.

Stephen pushed up a commit with updated `filter_sensitive_data` placeholder strings to prevent mangling  and refreshed the job.  Not all builds have completed yet, but in the ones that have, the tests still fail due to 403 response.

da70 commented:
Stephen and I are pairing today to debug this.  We deleted the cassette on my local and stepped through the tests to watch the cassette get created again.

The first `RestClient::Request.execute(...)` call is in the `fda_authenticate` method.  We stepped through until we got to the `Net::HTTP` request.  Somewhere along the way the user credentials he passes to  to `RestClient::Request.execute(...)` the `payload` key/value pair of `parameters` hash appeared to be getting lost.  The body of the POST request is empty.  `RestClient::Request::transmit` ends up raising a `RestClient::ServerBrokeConnection` error.  Tests return 500.
Kate checked the apache logs and verified that the POST request from my machine was empty, and that the server response was 500.

It could be that the payload is not being set properly in `RestClient::Request`.  Stephen and I are going to meet again this afternoon to debug some more.


Kate checked the apache logs

ekate commented:
Hi,

just FYI here is the code for DSpace REST client interaction written by one
of DSpace commiters, it uses rest_client gem:

https://github.com/peterdietz/dspace-rest-rails

when I was experimenting with Dspace/blacklight integration I was using
active_rest_client gem which uses farraday to perform the actual connection.

https://github.com/ekate/blacklight-dspace

Kate


On Mon, Jul 11, 2016 at 1:42 PM, da70 <notifications@github.com> wrote:

> Stephen and I are pairing today to debug this. We deleted the cassette on
> my local and stepped through the tests to watch the cassette get created
> again.
>
> The first RestClient::Request.execute(...) call is in the fda_authenticate
> method. We stepped through until we got to the Net::HTTP request.
> Somewhere along the way the user credentials he passes to to
> RestClient::Request.execute(...) the payload key/value pair of parameters
> hash appeared to be getting lost. The body of the POST request is empty.
> RestClient::Request::transmit ends up raising a
> RestClient::ServerBrokeConnection error. Tests return 500.
> Kate checked the apache logs and verified that the POST request from my
> machine was empty, and that the server response was 500.
>
> It could be that the payload is not being set properly in
> RestClient::Request. Stephen and I are going to meet again this afternoon
> to debug some more.
>
> Kate checked the apache logs
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NYULibraries/ichabod/pull/233#issuecomment-231808767>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACa_5vJqxAj7R6ozC6F3c3xXtrHj1rRtks5qUoCAgaJpZM4I2e0b>
> .
>



-- 
Ekaterina Pechekhonova
Senior Programmer/Analyst
New York University
Libraries
email: ekate@nyu.edu
phone: 212-992-9993

da70 commented:
I pushed up a new branch for debugging with a fully re-created cassette.  This branch works on my machine, but once again failed on Travis.
Kate looked through a different set of web logs this time and did find the Travis requests.  They were 403'ing, whereas the requests from my machine earlier when re-creating the cassette were getting 200's.

So the problem is Travis is making live requests.

eshadatta commented:
@NYULibraries/hydra ,

Is Travis making live requests for all or just for this? What if you tried using Faraday instead of Rest_Client? Just out of curiosity

da70 commented:
Rosie is not making live calls, we checked.  We haven't checked the others, but we're wondering if `git_geo_blacklight_reader_spec.rb` is making live calls as it takes so long and like FDA uses a token.  One way to test that would be to take the git repo offline or rename it and see if the tests on Travis fail.

We tried using Faraday very quickly yesterday but the results were inconclusive because it kept needed to re-create the cassette due to `fda_authenticate` not returning the right token.  I will fix that and see what happens on Travis.

Interestingly, our `spec_helper.rb` is specifying `:webmock` instead of `:faraday` for the library that intercepts HTTP requests:

```ruby
VCR.configure do |c|
...
  c.hook_into :webmock
...
end
```

When I changed this to `:faraday` we got a different kind of breakage for `fda_authenticate` that we'd temporarily switched to Faraday.  Possibly it is okay to specify `:webmock` because that replaces `Net::HTTP` with its own modified `Net::HTTP` object, and perhaps Faraday uses `Net::HTTP` -- haven't verified this, though.

eshadatta commented:
@da70 @sgbalogh ,

urg...so sorry this is so painful. It's ok to use webmock and faraday gem. I use the faraday gem to connect to Rosie and it's using the cassette(which has been using webmock), so I don't think you need to change the VCR.configure settings. If you could go ahead and change the gitgeoblacklight name, would it know to use the cassette? Just curious.

da70 commented:
The other possibility is that the specs on Travis are sending requests with signatures that differ slightly from what is registered via VCR.  This would mean that VCR goes in record mode and allows live requests for the purpose of updating the cassette.  A way to test for this on Travis would be to take checksum of the cassette file before and after tests run.

Stephen also pointed out that this code does not seem to be working in `application.rb`:

```ruby
# Don't run this initializer on travis.
Figs.load(stage: Rails.env) unless ENV['TRAVIS']
```

...since Travis is trying to clone configula repo and failing.

da70 commented:
The tricky part of checking gitgeoblacklight is that we don't have access to the server logs (Github), so can't do what Kate did with DSpace and check for live calls.  If we disable the repo, the Github API at the endpoint we use to load from should return an error response which would show up in the build.
If these errors happen in the build, chances are the cassette is not being used, or else it is loaded but the request signature is changing so VCR makes live calls in an attempt to update it.

We have verified that the `describe` block in the spec file has vcr configuration in memory.

eshadatta commented:
@da70 ,

Sure, I can disable the repo. How do I do that?

da70 commented:
@eshadatta: we basically need the endpoint "https://api.github.com/repos/NYULibraries/test_sdr_metadata/" to become invalid.  Maybe the easiest thing to do would be to rename the repo "test_sdr_metadata_DISABLED" or something like that.  Or if anyone has any other ideas, feel free to suggest.

This is probably not a 100% guarantee that we've disabled it since it's possible that the Github API code uses an internal ID for the repo instead of the name/path it exposes in the name of the endpoint.  I will trigger a build now to get a baseline.  After that if you would disable the repo, we can trigger another build.  I can also remove the VCR option from my local to make sure that the specs fail against the Github server.

I'll set up a branch for testing and trigger a build.  After it's done I'll let you know so you can disable the repo.

eshadatta commented:
@da70 , ok, will do. There might be about a 30 minute delay as I'm leaving Brooklyn soon to head into the office.

da70 commented:
@eshadatta: no rush at all.

da70 commented:
Baseline build has started: [feature/jv_fda_rest_reader_test-if-gitgeoblacklightreader-making-live-calls #1350](https://travis-ci.org/NYULibraries/ichabod/builds/144184205).

Branched off of Stephen's branch instead of `development` because Kate says `development` is not building right now.

da70 commented:
Despite what I wrote above, I did accidentally create a branch off `development`.  Deleted it and started over using Stephen's branch.

New baseline build: [#1351](https://travis-ci.org/NYULibraries/ichabod/builds/144186179)

da70 commented:
Previous one wasn't right.  Again...[#1352](https://travis-ci.org/NYULibraries/ichabod/builds/144189641)

da70 commented:
Travis and/or Github was getting confused about the tree.
Created a new branch.
Build: [#1356](https://travis-ci.org/NYULibraries/ichabod/builds/144191267)

da70 commented:
In my local `git_geo_blacklight_reader_spec.rb`, I removed the vcr arg from the '#read' `describe` block and ran the tests.  They failed with VCR errors:

```
VCR::Errors::UnhandledHTTPRequestError: 

================================================================================
An HTTP request has been made that VCR does not know how to handle:
  GET https://api.github.com/repos/NYULibraries/test_sdr_metadata/contents/

There is currently no cassette in use. There are a few ways
you can configure VCR to handle this request:

  * If you're surprised VCR is raising this error
    and want insight about how VCR attempted to handle the request,
    you can use the debug_logger configuration option to log more details [1].
  * If you want VCR to record this request and play it back during future test
    runs, you should wrap your test (or this portion of your test) in a
    `VCR.use_cassette` block [2].
  * If you only want VCR to handle requests made while a cassette is in use,
    configure `allow_http_connections_when_no_cassette = true`. VCR will
    ignore this request since it is made when there is no cassette [3].
  * If you want VCR to ignore this request (and others like it), you can
    set an `ignore_request` callback [4].

[1] https://www.relishapp.com/vcr/vcr/v/2-9-3/docs/configuration/debug-logging
[2] https://www.relishapp.com/vcr/vcr/v/2-9-3/docs/getting-started
[3] https://www.relishapp.com/vcr/vcr/v/2-9-3/docs/configuration/allow-http-connections-when-no-cassette
[4] https://www.relishapp.com/vcr/vcr/v/2-9-3/docs/configuration/ignore-request
================================================================================


./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:137:in `first_tree'
./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:107:in `get_download_urls'
./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:96:in `read_download_files'
./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:80:in `parse_download_files'
./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:51:in `map'
./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:12:in `read'
./spec/ichabod/resource_set/source_readers/git_geo_blacklight_reader_spec.rb:22:in `block (3 levels) in <module:SourceReaders>'
./spec/ichabod/resource_set/source_readers/git_geo_blacklight_reader_spec.rb:26:in `block (3 levels) in <module:SourceReaders>'
```

I did what the error message suggested and added this to `spec_helper.rb` VCR configuration:

```ruby
VCR.configure do |c|
...
  c.allow_http_connections_when_no_cassette = true
...
end
```

...and ran the tests again.  They completed successfully in 213.156493.
With cassette being used, the completion time is 56.675761.

This error is not showing up in Travis builds for Stephen's FDA branch.

da70 commented:
Build for 2.1.3 completed, all tests pass: [#1356.1 passed|https://travis-ci.org/NYULibraries/ichabod/jobs/144191268].

@eshadatta, feel free to disable the repo whenever you are ready.

da70 commented:
@eshadatta, please don't disable yet.  I noticed on my local the cassette had been rebuilt, and I think that's because Stephen had fixed the VCR placeholders.  It is likely that Travis had to rebuild the cassette, too.  I have pushed up the new cassettte from my local and started another Travis build for baseline: [#1357](https://travis-ci.org/NYULibraries/ichabod/builds/144207922)

da70 commented:
Huh.  [#1357](https://travis-ci.org/NYULibraries/ichabod/builds/144207922) failed with this error:

```Bash
 1) Ichabod::ResourceSet::SourceReaders::GitGeoBlacklightReader#read 
     Failure/Error: subject(:read) { reader.read }
     URI::InvalidURIError:
       bad URI(is not URI?): https://api.github.com/repos/7n8nnsizBM2dnW9M2>i49@/contents/
     # ./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:137:in `first_tree'
     # ./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:107:in `get_download_urls'
     # ./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:96:in `read_download_files'
     # ./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:80:in `parse_download_files'
     # ./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:51:in `map'
     # ./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:12:in `read'
     # ./spec/ichabod/resource_set/source_readers/git_geo_blacklight_reader_spec.rb:22:in `block (3 levels) in <module:SourceReaders>'
     # ./spec/ichabod/resource_set/source_readers/git_geo_blacklight_reader_spec.rb:23:in `block (3 levels) in <module:SourceReaders>'
```

It looks like `c.filter_sensitive_data('7n8nnsizBM2dnW9M2>i49@') { ENV['GIT_GEO_SPATIAL_MD_URL'] }` in `spec_helper.rb` did not work on Travis.  When I run the tests on the exact same commit on my local it passes:

```Bash
david@GURGI:ichabod/> git checkout 0031485
HEAD is now at 0031485... spatial_data_repository.yml was updated by running git_geo_blacklight_reader_spec.rb
david@GURGI:ichabod/> bundle exec rspec spec/ichabod/resource_set/source_readers/git_geo_blacklight_reader_spec.rb 
DEPRECATION WARNING: String based terminators are deprecated, please use a lambda. (called from included at /Users/david/.rvm/gems/ruby-2.1.3@ichabod/bundler/gems/authlogic-e4b2990d6282/lib/authlogic/session/callbacks.rb:66)
DEPRECATION WARNING: String based terminators are deprecated, please use a lambda. (called from included at /Users/david/.rvm/gems/ruby-2.1.3@ichabod/bundler/gems/authlogic-e4b2990d6282/lib/authlogic/session/callbacks.rb:67)
Run options: exclude {:half_baked=>true}
.......

Finished in 55.92 seconds
7 examples, 0 failures

Randomized with seed 17873

Coverage report generated for RSpec to /Users/david/Documents/programming/src/dlts/ichabod/coverage. 887 / 995 LOC (89.15%) covered.
Coverage report Rcov style generated for RSpec to /Users/david/Documents/programming/src/dlts/ichabod/coverage/rcov
[Coveralls] Outside the CI environment, not sending data.
```

da70 commented:
A working hypothesis: when Stephen initially tried to build, because of the confusion with the VCR `filter_sensitive_data` collisions on placeholders, the FDA cassette was getting invalidated by successive string replacements that increasingly mangled the filtered cassette values.  This forced VCR to make live calls, which failed due to absence of credentials on Travis.

After Stephen changed the placeholders to garbage strings generated with password creator program, this fixed collisions on local instances (mine at least), but introduced a bug that only happens on Travis -- `filter_sensitive_data` does not seem to be restoring the sensitive data on playback.  So once again, the cassette needs to be rebuilt during the rspec run.
It could be some of the characters in the garbage strings are throwing off the string replace, but only on certain machines, perhaps due to differing Rubies or dependencies (which would be weird given the Gemfile.lock).

In the case of GitGeoBlacklightReader spec, the tests can't even run because the sensitive data replacement is failing in the endpoint string, preventing Travis from even making a request to the right server.

da70 commented:
Changed branch `chore/test-if-gitgeoblacklightreader-making-live-calls`:
* Removed all rvm Ruby version except 2.1.3 from Travis file.
* Reverted VCR.configure filtering.  Removed FDA stuff and put in old placeholders.
```ruby
  c.filter_sensitive_data('user') { ENV['ICHABOD_ROSIE_USER'] }
  c.filter_sensitive_data('password') { ENV['ICHABOD_ROSIE_PASSWORD'] }
  c.filter_sensitive_data('user_token') { ENV['ICHABOD_GIT_USER_TOKEN'] }
  c.filter_sensitive_data('foo/bar') { ENV['GIT_GEO_SPATIAL_MD_URL'] }
```

Started a new build: [Build #1359](https://travis-ci.org/NYULibraries/ichabod/builds/144302515)

da70 commented:
Build failed because forgot to revert `spatial_data_repository.yml` cassette file with original placeholders.  Did so now, and started another build: [Build #1361](https://travis-ci.org/NYULibraries/ichabod/builds/144308123)

da70 commented:
[Build #1361](https://travis-ci.org/NYULibraries/ichabod/builds/144308123) passed.  This would appear to be a good baseline.  Tomorrow I'll ask Esha to disable the repo and we'll run this again to test that it is running against the cassette.

da70 commented:
Back in Stephen's branch: added code to the FDA rest reader spec to record the MD5 digest of the cassette file before and after the tests run.  Started a new build: [Build #1362](https://travis-ci.org/NYULibraries/ichabod/builds/144310096).

da70 commented:
**FDA**

[Build #1362](https://travis-ci.org/NYULibraries/ichabod/builds/144310096) failed.  The cassette file was changed in the course of the spec suite run.  Before the tests the MD5 digest was "c1aa70bba627bd70b3723ce07f133698" (as it is on my local), but after the tests had run, the digest was "7bc7fe5839de91908aefe4c0eb3af953".
On my local, the digest for the cassette file is "c1aa70bba627bd70b3723ce07f133698" both before and after tests run.

I didn't put in code to print the diff because if the difference has something to do with `filter_sensitive_data` not working right on Travis, there is the danger of sensitive data being printed in the Travis build output.

Given this and the accidental discovery in the `GitGeoBlacklightReader` experiment, I think it is possible that somehow the new `filter_sensitive_data` calls in Stephen's branch are not working correctly on Travis, which forces VCR to recreate the HTTP interactions in the cassette with each build.

I checked the raw Travis log to compare VCR and WebMock gem versions against my local.  The are the same.

I tried another build: [Build #1363](https://travis-ci.org/NYULibraries/ichabod/builds/144312107).  Same result.  The MD5 digests before and after do not match on Travis.

da70 commented:
**FDA**

Temporarily changed the VCR configuration to do the simplest filtering possible.  Removed all filtering except for FDA replacements, and changed placeholders to lowercase letters only, in case special characters were causing problems in string find/replace on Travis:

```ruby
  c.filter_sensitive_data('fdaurl') { ENV['FDA_REST_URL'] }
  c.filter_sensitive_data('fdauser') { ENV['FDA_REST_USER'] }
  c.filter_sensitive_data('fdapassword') { ENV['FDA_REST_PASS'] }
```

This didn't fix the problem: [Build #1365|https://travis-ci.org/NYULibraries/ichabod/builds/144471684]

MD5 digests for cassette file before and after tests run are different.

da70 commented:
Stephen and I will be working on this this afternoon.  We talked and agreed that it is a virtual impossibility that the `diff` will expose the credentials.  Travis doesn't have them, and the cassette was made on my computer which appears to be doing the filtering correctly, so the credentials should not be in the cassette file.

da70 commented:
**GitGeoBlacklightReader**

Esha disabled the repo.  On my local, did a check to see that the disabling worked:

* Ran tests, all succeeded (using cassette).
* Removed `vcr` option from the `describe` "#read" block, added `c.allow_http_connections_when_no_cassette = true` to `spec_helper.rb`, and ran tests.
  * Specs failed with errors:
```ruby
TypeError: no implicit conversion of String into Integer
./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:115:in `[]'
./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:115:in `block in get_download_urls'
./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:114:in `get_download_urls'
./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:96:in `read_download_files'
./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:80:in `parse_download_files'
./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:51:in `map'
./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:12:in `read'
./spec/ichabod/resource_set/source_readers/git_geo_blacklight_reader_spec.rb:22:in `block (3 levels) in <module:SourceReaders>'
./spec/ichabod/resource_set/source_readers/git_geo_blacklight_reader_spec.rb:23:in `block (3 levels) in <module:SourceReaders>'
```
...which is the result of `first_tree` not being able to get the data.  Stepping through to breakpoint at `...octokit-3.8.0/lib/octokit/client.rb` line 340 to see the response from the server: status was "301 Moved Permanently" from location https://api.github.com/repositories/36681519/contents/.  Interesting in that this appears to be a Github internal ID for our repo...but regardless the disabling worked.
* Re-enabled the cassette: the tests passed again.  Repeated a few times to be sure.

Pushed to Github and triggered a new build: [Build #1367](https://travis-ci.org/NYULibraries/ichabod/builds/144520039).

da70 commented:
**GitGeoBlacklightReader**

[Build #1367](https://travis-ci.org/NYULibraries/ichabod/builds/144520039) succeeded will all specs passing, presumably the cassette was used.

I pushed up a new [commit](https://github.com/NYULibraries/ichabod/commit/d3d39f6f7b3d5f9af527389bb919ac7115a69baf) with cassette disabled and live connection by VCR allowed: [Build #1368](https://travis-ci.org/NYULibraries/ichabod/builds/144523784).
This build failed, but with a different error than what I get on my local:

```Bash
Failures:
  1) Ichabod::ResourceSet::SourceReaders::GitGeoBlacklightReader#read 
     Failure/Error: subject(:read) { reader.read }
     Octokit::Unauthorized:
       GET https://api.github.com/repos/foo/bar/contents/: 401 - Bad credentials // See: https://developer.github.com/v3
     # ./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:137:in `first_tree'
     # ./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:107:in `get_download_urls'
     # ./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:96:in `read_download_files'
     # ./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:80:in `parse_download_files'
     # ./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:51:in `map'
     # ./lib/ichabod/resource_set/source_readers/git_geo_blacklight_reader.rb:12:in `read'
     # ./spec/ichabod/resource_set/source_readers/git_geo_blacklight_reader_spec.rb:22:in `block (3 levels) in <module:SourceReaders>'
     # ./spec/ichabod/resource_set/source_readers/git_geo_blacklight_reader_spec.rb:23:in `block (3 levels) in <module:SourceReaders>'
```

Note that the cassette placeholder "foo/bar" is present in the endpoint URL being used by `first_tree`.  This suggests that the cassette is being read, but not filtered, when in fact it is not supposed to be read in at all.  The endpoint should be the real one, not the one that was filtered when recorded to the cassette.
When I line debug the rspec run of the same commit on my local, I see that `first_tree` is using the correct, non-placeholder-substituted URL, which is why I get the 301 error locally instead of the 401 Travis is getting.

This suggests that Travis is loading the cassette but not filtering it.  I can't think of any other way for 'foo/bar' to be showing up like this.

da70 commented:
** GitGeoBlacklightReader**

Restarted the `development` build from 14 days ago: [Build #1319](https://travis-ci.org/NYULibraries/ichabod/builds/141140701).

This build doesn't contain any code from Stephen's branch or from my testing.

da70 commented:
Actually, it turns out that `git_geo_blacklight_reader_spec.rb` has "foo/bar" hardcoded:

```ruby
describe GitGeoBlacklightReader do
    let(:repo_url) { ENV['GIT_GEO_SPATIAL_MD_URL'] || 'foo/bar'}
    let(:access_token) { ENV['ICHABOD_GIT_USER_TOKEN'] || 'user_token' }
```

...and that's why Travis has it in the URL and my local doesn't, because Travis doesn't have the real URL from `ENV`.  Not sure what function this code is serving.  Maybe something to do with this:

```ruby
        before do
          allow(resource_set).to receive(:repo_url) { repo_url }
          allow(resource_set).to receive(:collection_code) { collection_code }
          allow(resource_set).to receive(:access_token) { access_token }
``` 
?

da70 commented:
**GitGeoBlacklightReader**

`development` build succeeded (with cassette, presumably): [Job #1319.1](https://travis-ci.org/NYULibraries/ichabod/jobs/141140702)

da70 commented:
**FDA**

[Build #1369](https://travis-ci.org/NYULibraries/ichabod/builds/144539617) failed.  The diff of before/after of the cassette:

```
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
2008a2009,2041
> - request:
>     method: post
>     uri: https://archive.nyu.edu/rest/login
>     body:
>       encoding: UTF-8
>       string: '{"email":"username","password":"password"}'
>     headers:
>       Accept:
>       - "*/*; q=0.5, application/xml"
>       Accept-Encoding:
>       - gzip, deflate
>       Content-Type:
>       - application/json
>       Content-Length:
>       - '42'
>       User-Agent:
>       - Ruby
>   response:
>     status:
>       code: 403
>       message: Forbidden
>     headers:
>       Date:
>       - Wed, 13 Jul 2016 19:02:00 GMT
>       Server:
>       - Apache-Coyote/1.1
>       Content-Length:
>       - '0'
>     body:
>       encoding: UTF-8
>       string: ''
>     http_version: 
>   recorded_at: Wed, 13 Jul 2016 19:02:00 GMT
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
```

da70 commented:
There is in `fda_rest_reader_spec.rb`, this code:

```ruby
        let(:fda_rest_user) { ENV['FDA_REST_USER'] || 'username' }
        let(:fda_rest_pass) { ENV['FDA_REST_PASS'] || 'password' }
```

...similar to what we found in `GitGeoBlacklightReader`.  We are going to try and remove the `||` stuff.

da70 commented:
*FDA*

Changed to:

```ruby
        let(:fda_rest_user) { ENV['FDA_REST_USER'] || 'fdauser' }
        let(:fda_rest_pass) { ENV['FDA_REST_PASS'] || 'fdapassword' }
```

...and regenerated cassette.

Pushed.  New build: [Build #1370](https://travis-ci.org/NYULibraries/ichabod/builds/144554576)

da70 commented:
**FDA**

[Build #1370](https://travis-ci.org/NYULibraries/ichabod/builds/144554576) failed.  The diff:

```
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
2008a2009,2041
> - request:
>     method: post
>     uri: https://archive.nyu.edu/rest/login
>     body:
>       encoding: UTF-8
>       string: '{"email":"fdauser","password":"fdapassword"}'
>     headers:
>       Accept:
>       - "*/*; q=0.5, application/xml"
>       Accept-Encoding:
>       - gzip, deflate
>       Content-Type:
>       - application/json
>       Content-Length:
>       - '44'
>       User-Agent:
>       - Ruby
>   response:
>     status:
>       code: 403
>       message: Forbidden
>     headers:
>       Date:
>       - Wed, 13 Jul 2016 19:31:27 GMT
>       Server:
>       - Apache-Coyote/1.1
>       Content-Length:
>       - '0'
>     body:
>       encoding: UTF-8
>       string: ''
>     http_version: 
>   recorded_at: Wed, 13 Jul 2016 19:31:27 GMT
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
```

da70 commented:
**FDA**

We might have solved it: [Build #1376](https://travis-ci.org/NYULibraries/ichabod/builds/144586952)

We removed `c.filter_sensitive_data` for the endpoint string so that URL would not be changed, then regenerated the cassette.  All tests passed.

Stephen is going to try this now on his original branch.

da70 commented:
**FDA**

We have fixed the problems with Stephen's FDA tests on Travis, however now Rosie test code and cassette need to be re-done as they are failing due to cassette placeholder changes.

Stephen has pushed up a change that should fix it.  We'll check the build results tomorrow: [Build #1383](https://travis-ci.org/NYULibraries/ichabod/builds/144604972).

da70 commented:
@sgbalogh, looks good!  Only failing cucumber tests for NYU Open Access Books.  I recall in earlier meetings people had mentioned that these were broken.  I updated the Solr index for this project recently so it could be that the Solr query used by Ichabod isn't returning the same first 5 records as before.  Probably we need make the data load query deterministic to avoid having to rewrite tests.  I'll ask at tech meeting today if anyone is working on this issue already.  If not, I can do it.

sgbalogh commented:
@da70: yep, I think we did it! :tada: Thanks for all the help. I still need to rebase, which I will try to do today.

da70 commented:
@sgbalogh -- sounds good.  Congrats!

The `development` branch tests have all passed: [Job #1319.1](https://travis-ci.org/NYULibraries/ichabod/jobs/141140702), which means GitGeoBlacklightReader is using the cassette properly because the Github repo with the data has been disabled.

@eshadatta, I think it's okay to re-enable the repo now.  Thanks much.

da70 commented:
@sgbalogh, the cucumber tests in `display.feature` do not fail on my local.  I ran them twice.  Looks like we still have a small mystery to solve.

da70 commented:
@sgbalogh, as you may have seen in the email discussion today, it seems likely that after you rebase on `development` this problem will disappear.  That branch contains updated cucumber tests and cassettes that your branch doesn't have.  Still don't know why the tests fail on Travis but not my local, but maybe we can leave that go since your branch is using live requests to Solr in its cucumber suite, and once you rebase, it will be using cassettes, which are static.

da70 commented:
Just to shed a little more light on the endpoint filtering issue: the reason that filter had to be removed from `spec_helper.rb` is because in `fda_rest_reader_spec.rb` there is this:

```ruby
        let(:fda_rest_url) { 'https://archive.nyu.edu/rest' }
```

...instead of this:

```ruby
        let(:fda_rest_url) { ENV['FDA_REST_URL'] | 'https://archive.nyu.edu/rest' }
```

...which led to a filtering asymmetry in the recording/playback across locals and Travis (configula-enabled vs. non-configula-enabled).

@sgbalogh, we might want to consider making the placeholder scheme you've used for Rosie as the standard.  So instead of this:

```ruby
  c.filter_sensitive_data('ichabod_rosie_user') { ENV['ICHABOD_ROSIE_USER'] }
  c.filter_sensitive_data('ichabod_rosie_password') { ENV['ICHABOD_ROSIE_PASSWORD'] }
  c.filter_sensitive_data('Ue7PHK9WtD824V8{6u>rBU') { ENV['ICHABOD_GIT_USER_TOKEN'] }
  c.filter_sensitive_data('7n8nnsizBM2dnW9M2>i49@') { ENV['GIT_GEO_SPATIAL_MD_URL'] }
  c.filter_sensitive_data('p4oWnb7r2H,X2scZ3ww76(') { ENV['FDA_REST_USER'] }
  c.filter_sensitive_data('2E4%dX;4B4hi4oHYJU4Lm2') { ENV['FDA_REST_PASS'] }
```

...we have this:

```ruby
  c.filter_sensitive_data('ichabod_rosie_user') { ENV['ICHABOD_ROSIE_USER'] }
  c.filter_sensitive_data('ichabod_rosie_password') { ENV['ICHABOD_ROSIE_PASSWORD'] }
  c.filter_sensitive_data('ichabod_git_user_token') { ENV['ICHABOD_GIT_USER_TOKEN'] }
  c.filter_sensitive_data('git_geo_spatial_md_url') { ENV['GIT_GEO_SPATIAL_MD_URL'] }
  c.filter_sensitive_data('fda_rest_user') { ENV['FDA_REST_USER'] }
  c.filter_sensitive_data('fda_rest_pass') { ENV['FDA_REST_PASS'] }
```

This automatically protects against collisions within the set of filters because `ENV` only has one value per variable name.  It doesn't protect as well against accidental collision within the filtered content, though, as the garbage strings do.  I like how readable the meaningful strings are, making them easier to find in the cassette, but that is not necessarily a priority.
Whichever we choose, it might be good to stay consistent to make things less confusing for posterity -- including us, months from now when we've forgotten how all this works.
Speaking of which, not necessarily in your branch, but we should probably put some comments in the code to document all this VCR stuff.  Or else put it in our developers' documentation.

sgbalogh commented:
@da70 agreed about the filtering convention recommendations –– I made the changes and am currently re-recording cassettes for GeoBlacklight and Rosie.

I have no real preference as to whether we document this in the code itself or in the developers' documentation. Which do you think might be more useful? At the very least, I could put a line in the VCR configuration section of `spec_helper.rb` that points to a more detailed discussion of filtering in the documentation?

da70 commented:
@sgbalogh, that's a good idea!  The code comment pointing to the doc.  There are multiple ways we could handle this VCR configuration stuff and we may need to iterate, so perhaps developer documentation is the best place to do that, as we could revise as needed without necessarily having to touch the code.

sgbalogh commented:
@da70: added a minimal comment to refer code users to see the Ichabod documentation. Assuming all passes again, let me know if there's anything else that could use sprucing up?

Also –– this probably doesn't matter –– but I was unable to re-record the geoblacklight cassette (hasn't been re-enabled on GitHub yet), so I just updated the filter strings via global find-and-replace on the old cassette.

eshadatta commented:
@NYULibraries/hydra ,

I've updated the private repo's name back to what it was for the sdr data. Just a fyi.

sgbalogh commented:
@eshadatta, thanks!
